File: EventOrganizer.cs
// EventOrganizer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class Task
{
    [JsonPropertyName("id")]
    public string Id { get; set; }
    [JsonPropertyName("description")]
    public string Description { get; set; }
    [JsonPropertyName("category")]
    public string Category { get; set; }
    [JsonPropertyName("completed")]
    public bool Completed { get; set; }

    public Task() { }
    public Task(string description, string category)
    {
        Id = Guid.NewGuid().ToString().Substring(0,8);
        Description = description;
        Category = category;
        Completed = false;
    }
}

class Event
{
    [JsonPropertyName("id")]
    public string Id { get; set; }
    [JsonPropertyName("name")]
    public string Name { get; set; }
    [JsonPropertyName("date")]
    public string Date { get; set; }
    [JsonPropertyName("description")]
    public string Description { get; set; }
    [JsonPropertyName("tasks")]
    public List<Task> Tasks { get; set; } = new List<Task>();

    public Event() { }
    public Event(string name, string date, string description)
    {
        Id = Guid.NewGuid().ToString().Substring(0,8);
        Name = name;
        Date = date;
        Description = description;
    }
}

class Organizer
{
    private List<Event> events = new List<Event>();
    private readonly string dataFile = "events.json";
    private readonly JsonSerializerOptions options = new JsonSerializerOptions { WriteIndented = true };

    public Organizer() => Load();

    private void Load()
    {
        if (!File.Exists(dataFile)) return;
        string json = File.ReadAllText(dataFile);
        events = JsonSerializer.Deserialize<List<Event>>(json) ?? new List<Event>();
    }

    private void Save()
    {
        string json = JsonSerializer.Serialize(events, options);
        File.WriteAllText(dataFile, json);
    }

    private Event GetEvent(string id) => events.FirstOrDefault(e => e.Id == id);

    public void Create(string name, string date, string desc)
    {
        var e = new Event(name, date, desc);
        events.Add(e);
        Save();
        Console.WriteLine($"✅ Event created: {e.Name} (ID: {e.Id})");
    }

    public void AddTask(string eventId, string desc, string category)
    {
        var e = GetEvent(eventId);
        if (e == null)
        {
            Console.WriteLine($"Event {eventId} not found.");
            return;
        }
        var t = new Task(desc, category);
        e.Tasks.Add(t);
        Save();
        Console.WriteLine($"✅ Task added: {t.Description} (ID: {t.Id})");
    }

    public void List()
    {
        if (!events.Any())
        {
            Console.WriteLine("No events.");
            return;
        }
        Console.WriteLine("\n📋 Events:");
        foreach (var e in events)
        {
            int total = e.Tasks.Count;
            int done = e.Tasks.Count(t => t.Completed);
            int pct = total > 0 ? (int)(done * 100.0 / total) : 0;
            Console.WriteLine($"  {e.Name} (ID: {e.Id}) – {done}/{total} tasks done ({pct}%)");
        }
    }

    public void Show(string eventId)
    {
        var e = GetEvent(eventId);
        if (e == null)
        {
            Console.WriteLine($"Event {eventId} not found.");
            return;
        }
        int total = e.Tasks.Count;
        int done = e.Tasks.Count(t => t.Completed);
        int pct = total > 0 ? (int)(done * 100.0 / total) : 0;
        Console.WriteLine($"\n📋 Event: {e.Name}");
        if (!string.IsNullOrEmpty(e.Date)) Console.WriteLine($"Date: {e.Date}");
        if (!string.IsNullOrEmpty(e.Description)) Console.WriteLine($"Description: {e.Description}");
        Console.WriteLine($"Progress: {done}/{total} tasks done ({pct}%)");
        if (total == 0)
        {
            Console.WriteLine("No tasks yet.");
            return;
        }
        Console.WriteLine("\nTasks:");
        int i = 1;
        foreach (var t in e.Tasks)
        {
            string status = t.Completed ? "[X]" : "[ ]";
            string cat = !string.IsNullOrEmpty(t.Category) ? $" ({t.Category})" : "";
            Console.WriteLine($"  {i++}. {status} {t.Description}{cat} (ID: {t.Id})");
        }
    }

    public void MarkDone(string eventId, string taskId)
    {
        var e = GetEvent(eventId);
        if (e == null)
        {
            Console.WriteLine($"Event {eventId} not found.");
            return;
        }
        var t = e.Tasks.FirstOrDefault(task => task.Id == taskId);
        if (t == null)
        {
            Console.WriteLine($"Task {taskId} not found in this event.");
            return;
        }
        if (t.Completed)
        {
            Console.WriteLine("Task already completed.");
        }
        else
        {
            t.Completed = true;
            Save();
            Console.WriteLine($"✅ Task marked as done: {t.Description}");
        }
    }

    public void Delete(string eventId)
    {
        var e = GetEvent(eventId);
        if (e == null)
        {
            Console.WriteLine($"Event {eventId} not found.");
            return;
        }
        string name = e.Name;
        events.Remove(e);
        Save();
        Console.WriteLine($"✅ Event deleted: {name}");
    }

    static void Main(string[] args)
    {
        if (args.Length < 1)
        {
            Console.WriteLine("Usage: EventOrganizer <command> [options]");
            return;
        }
        var app = new Organizer();
        var cmd = args[0];
        var parsed = ParseArgs(args);
        switch (cmd)
        {
            case "create":
                if (args.Length < 2) { Console.WriteLine("create <name> [--date DATE] [--desc TEXT]"); return; }
                app.Create(args[1], parsed.GetValueOrDefault("date", ""), parsed.GetValueOrDefault("desc", ""));
                break;
            case "add-task":
                if (args.Length < 4) { Console.WriteLine("add-task <event_id> <description> [--category CAT]"); return; }
                app.AddTask(args[1], args[2], parsed.GetValueOrDefault("category", ""));
                break;
            case "list":
                app.List();
                break;
            case "show":
                if (args.Length < 2) { Console.WriteLine("show <event_id>"); return; }
                app.Show(args[1]);
                break;
            case "done":
                if (args.Length < 3) { Console.WriteLine("done <event_id> <task_id>"); return; }
                app.MarkDone(args[1], args[2]);
                break;
            case "delete":
                if (args.Length < 2) { Console.WriteLine("delete <event_id>"); return; }
                app.Delete(args[1]);
                break;
            default:
                Console.WriteLine("Unknown command.");
                break;
        }
    }

    static Dictionary<string, string> ParseArgs(string[] args)
    {
        var dict = new Dictionary<string, string>();
        for (int i = 1; i < args.Length; i++)
        {
            if (args[i].StartsWith("--") && i + 1 < args.Length)
            {
                dict[args[i].Substring(2)] = args[++i];
            }
        }
        return dict;
    }
}

File: event_organizer.py
# event_organizer.py
import json
import os
import sys
import argparse
import uuid
from datetime import datetime

DATA_FILE = "events.json"

class Task:
    def __init__(self, description, category="", completed=False, task_id=None):
        self.id = task_id or str(uuid.uuid4())[:8]
        self.description = description
        self.category = category
        self.completed = completed

    def to_dict(self):
        return {"id": self.id, "description": self.description,
                "category": self.category, "completed": self.completed}

    @classmethod
    def from_dict(cls, data):
        return cls(data["description"], data.get("category", ""),
                   data.get("completed", False), data.get("id"))

class Event:
    def __init__(self, name, date="", description="", event_id=None):
        self.id = event_id or str(uuid.uuid4())[:8]
        self.name = name
        self.date = date
        self.description = description
        self.tasks = []

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "date": self.date,
            "description": self.description,
            "tasks": [t.to_dict() for t in self.tasks]
        }

    @classmethod
    def from_dict(cls, data):
        ev = cls(data["name"], data.get("date", ""), data.get("description", ""), data.get("id"))
        ev.tasks = [Task.from_dict(t) for t in data.get("tasks", [])]
        return ev

class Organizer:
    def __init__(self):
        self.events = []
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                data = json.load(f)
                self.events = [Event.from_dict(e) for e in data]

    def save(self):
        with open(DATA_FILE, "w") as f:
            json.dump([e.to_dict() for e in self.events], f, indent=2)

    def get_event(self, event_id):
        for e in self.events:
            if e.id == event_id:
                return e
        return None

    def create(self, name, date="", description=""):
        ev = Event(name, date, description)
        self.events.append(ev)
        self.save()
        print(f"✅ Event created: {ev.name} (ID: {ev.id})")

    def add_task(self, event_id, description, category=""):
        ev = self.get_event(event_id)
        if not ev:
            print(f"Event {event_id} not found.")
            return
        task = Task(description, category)
        ev.tasks.append(task)
        self.save()
        print(f"✅ Task added: {task.description} (ID: {task.id})")

    def list_events(self):
        if not self.events:
            print("No events.")
            return
        print("\n📋 Events:")
        for e in self.events:
            total = len(e.tasks)
            done = sum(1 for t in e.tasks if t.completed)
            pct = (done / total * 100) if total else 0
            print(f"  {e.name} (ID: {e.id}) – {done}/{total} tasks done ({pct:.0f}%)")

    def show(self, event_id):
        ev = self.get_event(event_id)
        if not ev:
            print(f"Event {event_id} not found.")
            return
        total = len(ev.tasks)
        done = sum(1 for t in ev.tasks if t.completed)
        pct = (done / total * 100) if total else 0
        print(f"\n📋 Event: {ev.name}")
        if ev.date:
            print(f"Date: {ev.date}")
        if ev.description:
            print(f"Description: {ev.description}")
        print(f"Progress: {done}/{total} tasks done ({pct:.0f}%)\n")
        if not ev.tasks:
            print("No tasks yet.")
            return
        print("Tasks:")
        for i, task in enumerate(ev.tasks, 1):
            status = "[X]" if task.completed else "[ ]"
            cat = f" ({task.category})" if task.category else ""
            print(f"  {i}. {status} {task.description}{cat} (ID: {task.id})")

    def mark_done(self, event_id, task_id):
        ev = self.get_event(event_id)
        if not ev:
            print(f"Event {event_id} not found.")
            return
        for task in ev.tasks:
            if task.id == task_id:
                if task.completed:
                    print(f"Task already completed.")
                else:
                    task.completed = True
                    self.save()
                    print(f"✅ Task marked as done: {task.description}")
                return
        print(f"Task {task_id} not found in this event.")

    def delete(self, event_id):
        ev = self.get_event(event_id)
        if not ev:
            print(f"Event {event_id} not found.")
            return
        self.events.remove(ev)
        self.save()
        print(f"✅ Event deleted: {ev.name}")

def main():
    parser = argparse.ArgumentParser(description="Event Organizer")
    subparsers = parser.add_subparsers(dest="cmd", required=True)

    create_parser = subparsers.add_parser("create")
    create_parser.add_argument("name")
    create_parser.add_argument("--date", help="YYYY-MM-DD")
    create_parser.add_argument("--desc", help="Description")

    add_task_parser = subparsers.add_parser("add-task")
    add_task_parser.add_argument("event_id")
    add_task_parser.add_argument("description")
    add_task_parser.add_argument("--category", default="")

    subparsers.add_parser("list")

    show_parser = subparsers.add_parser("show")
    show_parser.add_argument("event_id")

    done_parser = subparsers.add_parser("done")
    done_parser.add_argument("event_id")
    done_parser.add_argument("task_id")

    delete_parser = subparsers.add_parser("delete")
    delete_parser.add_argument("event_id")

    args = parser.parse_args()
    app = Organizer()

    if args.cmd == "create":
        app.create(args.name, args.date or "", args.desc or "")
    elif args.cmd == "add-task":
        app.add_task(args.event_id, args.description, args.category)
    elif args.cmd == "list":
        app.list_events()
    elif args.cmd == "show":
        app.show(args.event_id)
    elif args.cmd == "done":
        app.mark_done(args.event_id, args.task_id)
    elif args.cmd == "delete":
        app.delete(args.event_id)

if __name__ == "__main__":
    main()
